### PR TITLE
8273381: Assert in PtrQueueBufferAllocatorTest.stress_free_list_allocator_vm

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this fix of the test, eliminating a race that could
unexpectedly leave buffers in the "completed" list after all the threads
have shut down.

Testing:
mach5 tier1

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273381](https://bugs.openjdk.java.net/browse/JDK-8273381): Assert in PtrQueueBufferAllocatorTest.stress_free_list_allocator_vm

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5820/head:pull/5820` \
`$ git checkout pull/5820`

Update a local copy of the PR: \
`$ git checkout pull/5820` \
`$ git pull https://git.openjdk.java.net/jdk pull/5820/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5820`

View PR using the GUI difftool: \
`$ git pr show -t 5820`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5820.diff">https://git.openjdk.java.net/jdk/pull/5820.diff</a>

</details>

### CODE REVIEW

**Review by kstefanj — APPROVED** [review]
Lgtm

File: test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp [hunk 1]
```diff
@@ -180,12 +180,18 @@ class BufferNode::TestSupport::ProcessorThread : public JavaTestThread {
...
+    bool shutdown_requested = false;
     while (true) {
       BufferNode* node = _cbl->pop();
       if (node != NULL) {
         _allocator->release(node);
-      } else if (!Atomic::load_acquire(_continue_running)) {
+      } else if (shutdown_requested) {
         return;
+      } else if (!Atomic::load_acquire(_continue_running)) {
+        // To avoid a race that could leave buffers in the list after this
```
Comment (tschatzl) [inline test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp:191]: s/avoid/handle/ (or something similar).

The change does not eliminate (or avoids) the race, but handles the (not logic-breaking) results of the race.

**Review by tschatzl — APPROVED** [review]
Lgtm apart from the mentioned (imho) comment inaccuracy.

File: test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp (hunk 1 above)
Comment (tschatzl) [reply test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp:191]:
```suggestion
        // To handle a benign race that could leave buffers in the list after this
```

Comment (kimbarrett) [reply test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp:191]: The race involved is not a benign data race, it's a logic race that's not at all benign.  I could describe the race in more detail.

Comment (tschatzl) [reply test/hotspot/gtest/gc/shared/test_ptrQueueBufferAllocator.cpp:191]: I meant "benign" wrt to what is actually tested (functionality of the ptrqueuebuffer), but you are right about that my suggestion is the wrong wording. Just leave it as is then.

Thomas

**Pull request closed without merging**